pkg/webhook/storageclass.go

```diff
@@ -0,0 +1,174 @@
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		oldSc := &storagev1.StorageClass{}
+		if _, _, err := deserializer.Decode(oldRaw, nil, oldSc); err != nil {
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		return decideV1StorageClass(sc, oldSc, isUpdate)
+	default:
+		err := fmt.Errorf("expect resource to be %s", StorageClassV1GVR)
+		klog.Error(err)
+		return toV1AdmissionResponse(err)
+	}
+}
+
+func decideV1StorageClass(sc, oldsc *storagev1.StorageClass, isUpdate bool) *v1.AdmissionResponse {
+	reviewResponse := &v1.AdmissionResponse{
+		Allowed: true,
+		Result:  &metav1.Status{},
+	}
+
+	if isUpdate {
+		// TBD
+		return reviewResponse
+	}
+
+	if err := ValidateV1StorageClass(sc); err != nil {
+		reviewResponse.Allowed = false
+		reviewResponse.Result.Message = err.Error()
+	}
+	return reviewResponse
+}
+
+func ValidateV1StorageClass(sc *storagev1.StorageClass) error {
+	if sc == nil {
+		return fmt.Errorf("StorageClass is nil")
+	}
+
+	if sc.Provisioner != FilestoreCSIDriver {
+		return nil
+	}
+
+	params := sc.Parameters
+	if params == nil {
+		return nil
+	}
+
+	if _, ok := params["multishare"]; ok {
+		klog.Infof("Multishare Filestore CSI storage class detected")
+	}
+
+	// TBD: Fill the validation logic here.
+	return nil
+}
+
+func toV1AdmissionResponse(err error) *v1.AdmissionResponse {
+	return &v1.AdmissionResponse{
+		Result: &metav1.Status{
+			Message: err.Error(),
+		},
+	}
+}
+func mutateStorageClass(ar v1.AdmissionReview) *v1.AdmissionResponse {
+	klog.Info("mutating storageClass")
+	reviewResponse := &v1.AdmissionResponse{
+		Allowed: true,
+		Result:  &metav1.Status{},
+	}
+
+	if !(ar.Request.Operation == v1.Update || ar.Request.Operation == v1.Create) {
+		return reviewResponse
+	}
+
+	raw := ar.Request.Object.Raw
+	oldRaw := ar.Request.OldObject.Raw
+
+	deserializer := codecs.UniversalDeserializer()
+	switch ar.Request.Resource {
+	case StorageClassV1GVR:
+		sc := &storagev1.StorageClass{}
+		if _, _, err := deserializer.Decode(raw, nil, sc); err != nil {
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		oldSc := &storagev1.StorageClass{}
+		if _, _, err := deserializer.Decode(oldRaw, nil, oldSc); err != nil {
+			klog.Error(err)
+			return toV1AdmissionResponse(err)
+		}
+		klog.Infof("check patch for storageClass %s", sc.Name)
+		return applyV1StorageClassPatch(sc)
+	default:
+		err := fmt.Errorf("expect resource to be %s", StorageClassV1GVR)
+		klog.Error(err)
+		return toV1AdmissionResponse(err)
+	}
+}
+
+func applyV1StorageClassPatch(sc *storagev1.StorageClass) *v1.AdmissionResponse {
+	reviewResponse := &v1.AdmissionResponse{
+		Allowed: true,
+	}
+
+	if sc.Provisioner != FilestoreCSIDriver {
+		return reviewResponse
+	}
+
+	isMultishare, ok := sc.Parameters["multishare"]
+	if !ok || isMultishare == "false" || isMultishare == "False" {
+		return reviewResponse
+	}
+
+	scPatch := fmt.Sprintf(`[{"op":"add", "path":"/parameters/instancePrefix","value": "%s"}]`, sc.Name)
```

Comment:
what happens if instancePrefix is already there?

Reply:
As discussed, if the user sets an instanceprefix it will be validated for proper length only. Will submit those changes as a different PR with unit tests